### PR TITLE
34 add generic parsing logic

### PR DESCRIPTION
**Pull request by TNO-SlaFleur**
1. Added a parser which can find out the message type automatically and parse the message accordingly. Instead of doing all kinds of reflective black magic, I went for generating the mapping from message type to the class which should be used for parsing. By generating this mapping the result is fast and simple and regenerating should be breeze.

2. Allow to also parse from a dict: `ValidateValuesMixin.from_dict()`

3. Add in and fix lint and typecheck to ci and tox while also fixing any outstanding issues.

### CODE REVIEW

File: src/s2python/s2_parser.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                message_dict,
+                f"Unable to parse {message_type} as an S2 message. Type unknown.",
+            )
+
+        return TYPE_TO_MESSAGE_CLASS[message_type].parse_obj(message_str)
+
+    def parse_dict_as_any_message(self, message_dict: dict) -> S2Message:
+        message_type = self.get_message_type_from_dict(message_dict)
+
+        if message_type not in TYPE_TO_MESSAGE_CLASS:
```
Comment (victorgarcia98) [inline src/s2python/s2_parser.py:59]: Nice!

This also covers for the case of not having the `message_type` key.

File: src/s2python/s2_parser.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            )
+
+        return TYPE_TO_MESSAGE_CLASS[message_type].parse_obj(message_dict)
+
+    def parse_str_as_message(self, message: str, as_message: S2Message[C]) -> C:
+        """Parse the message to a specific S2 python message.
+
+        :param message: The message as a JSON-formatted string.
+        :param as_message: The type of message that is expected within the `message`
+        :raises: S2ValidationError
```
Comment (victorgarcia98) [inline src/s2python/s2_parser.py:72]: This could also raise a `json.JSONDecodeError`.

File: src/s2python/s2_parser.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    "PowerMeasurement": PowerMeasurement,
+    "ReceptionStatus": ReceptionStatus,
+    "ResourceManagerDetails": ResourceManagerDetails,
+    "RevokeObject": RevokeObject,
+    "SelectControlType": SelectControlType,
+    "SessionRequest": SessionRequest,
+}
+
+
+class S2Parser:
```
Comment (victorgarcia98) [inline src/s2python/s2_parser.py:40]: Maybe, we could have a method that accepts both a str and a dict:

```python
def parse(self, message: str | dict) -> S2Message:
    if isintance(message, str):
        message = json.loads(message)
    
    message_type = self.get_message_type_from_dict(message)

    if message_type not in TYPE_TO_MESSAGE_CLASS:
        raise S2ValidationError(
            message,
            f"Unable to parse {message_type} as an S2 message. Type unknown.",
        )

    return TYPE_TO_MESSAGE_CLASS[message_type].parse_obj(message)    
```

**Review by victorgarcia98 — COMMENTED** [review]
Nice feature! This is very useful to start creating objects from serialized data :smile:

**Comment by TNO-SlaFleur** [comment]
@victorgarcia98 Apologies, I kinda went overboard. I added in lint and typecheck as well as I ran into some issues actually running the tools. I fixed all outstanding issues and turned on the tools in ci as well. All of these changes got stuck into this PR. Hope you don't mind, otherwise I can look into splitting them into 2. I did take care to change the original description of the PR to include these changes.

File: src/s2python/s2_parser.py (hunk 3 above)
Comment (TNO-SlaFleur) [reply src/s2python/s2_parser.py:60]: Fixed!

File: src/s2python/s2_parser.py (hunk 2 above)
Comment (TNO-SlaFleur) [reply src/s2python/s2_parser.py]: Fixed!

**Review by victorgarcia98 — APPROVED** [review]
Nice, this is very handy!

Also, thanks for fixing the typing issues! :smile:

**Pull request merged**